### DISCUSSION TITLE
HELICS on systems with less common CPUs and OSes

### DISCUSSION THREAD

**Discussion by nightlark** [GMLC-TDC/HELICS#2325] [General]
I've tried building HELICS and running a simple local test involving a single echo federate connecting to a HELICS broker on a variety of machines with less common CPU architectures and/or operating systems in the [GCC Compile Farm](https://cfarm.tetaneutral.net/machines/list/#). Here are the results:

Worked:
- Apple Mac Mini M1 (aarch64) running Debian bookworm/sid Linux
- Lenovo HR350A with an ARMv8 Ampere eMag CPU (aarch64) running CentOS 8
- qemu virtual machine emulating a RISC-V processor, running Debian (note: warnings about casts from std::byte* to uint* increasing required alignment of target type in ActionMessage)
- SPARC T3-2 with an UltraSparc T3 (Niagara3) (sparc64) CPU, running Debian bookworm/sid

Failed:
- Physical OpenBSD server (amd64) running OpenBSD 7.0 (failed due to trying to link with -lrt)
- EdgeRouter4 (mips64) running Debian 10.11 buster (undefined reference to `__atomic_store_8' when linking the helics_broker; also warnings about casts from std::byte* to uint* increasing required alignment of target type in ActionMessage)
- SPARC64-VII system (sun4u CPU) running Solaris 11.3 (has gcc 5.5.0, need to get/build a newer version)
- IBM POWER8 (chrp CPU architecture) running AIX 7 2 (asio fails to compile with message saying it only supports Windows + POSIX; also needed to build cmake from source)

I haven't tried to fix the failed builds; the OpenBSD one is probably the easiest to get working (CMake try_compile to check if -lrt is needed on a system), the mips64 issue could possibly be linking with -latomic, and the rest would be more involved.

**Comment by phlptp** [comment]
I am going to try to fix the alignment warnings someday,  we see those on other platforms as well.  

I seem to recall the rt linking issue on other platforms as well,  I agree that one should be pretty straightforward as long as we can detect it in CMake

 IBM might see if you can build with asio disabled.  Is ZMQ available on those platforms?

**Reply by nightlark** [reply]
For all the builds that worked, I compiled using ZeroMQ as a subproject. On OpenBSD I turned off ZeroMQ due to compilation errors in the library -- the `-lrt` related error is later on when linking libhelics.

I just tried building with ASIO disabled on the IBM AIX system, building ZMQ printed many error messages (no ZMQ_IOTHEAD_POLLER_USE_* macro defined, various types not being declared that may be related to the undefined macro). With both ASIO and ZMQ disabled, it fails on `#include <ifaddrs.h>` in netif code.

Getting HELICS working with any useful cores enabled seems like it would involve patches to multiple dependencies AIX. It looks like AIX might be like BSD and just require including `sys/socket.h` to get the functions that Linux has in `ifaddrs.h`. It actually looks like Solaris 11 adds `ifaddrs.h`, and Solaris 10 requires a completely different approach to getting network interfaces.

There are a few PowerPC systems in the compile farm with some common Linux distros installed that I'll try at some point, and should be easier to get working than AIX.

**Comment by nightlark** [comment]
Some details on the specific gcc cfarm machine hostnames used, for future reference:
* gcc103 (Asahi Linux)
* gcc85 (Ampere ARM)
* gcc402 (RISC-V virtual machine)
* gcc102 (UltraSparc T3)
* gcc220 (OpenBSD)
* gcc119 (AIX)
